# How to read 10X Cellranger "Ranked Barcode Plot"
Below is showing a **Ranked Barcode Plot**, obtained after running 10X Genomics Cellranger's `cellranger count` function. 

<p align="center">
  <img width="50%" height="50%" src="RankedBarcodePlot.png">
</p>

- **y-axis**: number of UMI counts mapped to each barcode.

- **x-axis**: number of barcodes below "that" UMI counts. 

- The steep drop-off in plot seperates cell-associated barcodes (on the left of x-axis) and empty partitions (on the right of x-axis). 

- Barcodes can be determined to be cell-associated based on their UMI count or by their RNA-profile. In other words, if there is enough UMI counts or RNAs associated with the barcode, this barcode is cell-associated. 

# How to interpret the "Fraction Reads in Cells" metric?
[How to interpret the "Fraction Reads in Cells" metric?](https://kb.10xgenomics.com/hc/en-us/articles/360003919491-How-to-interpret-the-Fraction-Reads-in-Cells-metric-)
[Gene Expression Algorithms Overview](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview)

**Fraction Reads in Cells**
The fraction of valid-barcode, confidently-mapped-to-transcriptome reads with cell-associated barcodes.

Question: I see a low value for the "Fraction Reads in Cells". How can I interpret this metric?

Answer: A low "Fraction Reads in Cells" value is typically explained by the following:

1) High ambient RNA (background) in your sample. This ambient RNA comes from lysed/dead cells in your sample. Cell Ranger is able to confidently align the reads from ambient RNA to the transcriptome but the reads are not associated with a valid cell-containing GEM.

2) The cell-calling heuristic did not apply. For example, there may be higher variation in RNA content than expected (more cells with lower RNA content). The current cell-calling heuristic assumes a ten-fold variation in RNA content.

**Cell Ranger's algorithm for partitioning barcodes as cells versus background is based on the idea that barcodes for cells should have distinctly more transcript counts associated with them than the background barcodes.** This can be visualized by the ranked barcode plot in the web_summary.html file. More details on the cell filtering algorithm can be found here.

If you suspect that Cell Ranger's cell calling algorithm did not work well for your sample, please re-run cellranger count again or cellranger reanalyze with --force-cells option to call the expected number of cells.

**The flow of FASTQ files:**
- bcl files are converted to FASTQ files, this step generates FASTQ files with valid Index + Adapter sequences, and Undetermined FASTQ (sequences without valid Index + Adaptoer sequences).
- the determined FASTQ files will be mapped the the genome to generate count matrix. Ambient RNA (with Index + Adapter sequence) will also be mapped to the genome. But, these ambient RNA (each of them should have a barcode) are not classified as cell-associated. 
- whether or not a barcode (each barcode could have multiple transcripts/reads associated to it) is classified as cell-associated is determined by the cell-calling algorithm. 

**Mapping**

By default, reads that are transcriptomic (blue) are carried forward to UMI counting. In certain cases, such as when the input to the assay consists of nuclei, there may be high levels of intronic reads generated by unspliced transcripts. In order to count these intronic reads, the cellranger count and cellranger multi pipelines can be run with the option include-introns. If this option is used, any reads that map in the sense orientation to a single gene - which include the reads labeled transcriptomic (blue), exonic (light blue), and intronic (red) in the diagram above - are carried forward to UMI counting.

Furthermore, a read is considered uniquely mapping if it is compatible with only a single gene. Only uniquely mapping reads are carried forward to UMI counting.

**Note, in the Web Summary HTML, the set of reads carried forward to UMI counting is referred to as "Reads mapped confidently to transcriptome"**

**UMI counting**
- Reads that were confidently mapped to the transcriptome are placed into groups that share the same barcode, UMI, and gene annotation. If two groups of reads have the same barcode and gene, but their UMIs differ by a single base (i.e., are Hamming distance 1 apart), then one of the UMIs was likely introduced by a substitution error in sequencing. In this case, the UMI of the less-supported read group is corrected to the UMI with higher support.
- Cell Ranger again groups the reads by barcode, UMI (possibly corrected), and gene annotation. If two or more groups of reads have the same barcode and UMI, but different gene annotations, the gene annotation with the most supporting reads is kept for UMI counting, and the other read groups are discarded. In case of a tie for maximal read support, all read groups are discarded, as the gene cannot be confidently assigned.
- After these two filtering steps, each observed barcode, UMI, gene combination is recorded as a UMI count in the unfiltered feature-barcode matrix.

**Calling Cell Barcodes**
The algorithm has two key steps:

1. It uses a cutoff based on total UMI counts of each barcode to identify cells. This step identifies the primary mode of high RNA content cells.
2. Then the algorithm uses the RNA profile of each remaining barcode to determine if it is an “empty" or a cell containing partition. This second step captures low RNA content cells whose total UMI counts may be similar to empty GEMs.

In the first step, the original Cell Ranger cell calling algorithm is used to identify the primary mode of high RNA content cells, using a cutoff based on the total UMI count for each barcode. Cell Ranger takes as input the expected number of recovered cells, N (see --expect-cells). Let m be the 99th percentile of the top N barcodes by total UMI counts. All barcodes whose total UMI counts exceed m/10 are called as cells in the first pass.

**In the second step, a set of barcodes with low UMI counts that likely represent ‘empty’ GEM partitions is selected.** A model of the RNA profile of selected barcodes is created. This model, called the background model, is a multinomial distribution over genes. It uses Simple Good-Turing smoothing to provide a non-zero model estimate for genes that were not observed in the representative empty GEM set. Finally, the RNA profile of each barcode not called as a cell in the first step is compared to the background model. Barcodes whose RNA profile strongly disagrees with the background model are added to the set of positive cell calls. This second step identifies cells that are clearly distinguishable from the profile of empty GEMs, even though they may have much lower RNA content than the largest cells in the experiment.


